source/opt/type_manager.h

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#ifndef LIBSPIRV_OPT_TYPE_MANAGER_H_
+#define LIBSPIRV_OPT_TYPE_MANAGER_H_
+
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "module.h"
+#include "types.h"
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+// A class for managing the SPIR-V type hierarchy.
+class TypeManager {
+ public:
+  using IdToTypeMap = std::unordered_map<uint32_t, std::unique_ptr<Type>>;
+  using ForwardPointerVector = std::vector<std::unique_ptr<ForwardPointer>>;
+
+  TypeManager() = default;
+  TypeManager(const TypeManager&) = delete;
+  TypeManager(TypeManager&&) = delete;
+  TypeManager& operator=(const TypeManager&) = delete;
+  TypeManager& operator=(TypeManager&&) = delete;
+
+  // Returns the type for the given type |id|. Returns nullptr if the given |id|
+  // does not define a type.
+  Type* GetType(uint32_t id) const;
+  // Returns the number of types hold in this manager.
+  size_t NumTypes() const { return id_to_type_.size(); }
+
+  // Returns the forward pointer type at the given |index|.
+  ForwardPointer* GetForwardPointer(uint32_t index) const;
+  // Returns the number of forward pointer types hold in this manager.
+  size_t NumForwardPointers() const { return forward_pointers_.size(); }
+
+  // Analyzes the types and decorations on types in the given |module|.
+  void AnalyzeType(const spvtools::ir::Module& module);
```

Comment:
What if we call it a second time, on a different module, or on a modified version of the module?